sys/vfs_util/vfs_util.c

```diff
@@ -154,3 +154,136 @@ int vfs_file_sha256(const char* file, void *digest,
 
 }
 #endif /* MODULE_HASHES */
+
+int vfs_is_dir(const char *path)
+{
+    assert(path);
+
+    int err;
+    struct stat stat;
+    if (*path != '/') {
+        /* only accept absolute paths */
+        return -EINVAL;
+    }
+    if ((err = vfs_stat(path, &stat)) < 0) {
+        return err;
+    }
+    return ((stat.st_mode & S_IFMT) == S_IFDIR);
+}
+
+/**
+ * @brief   Removes additional "/" slashes from @p path
+ *
+ * @param[in]   path    Path to be prepared
+ */
+static void _vfs_prepare_path(char *path)
+{
+    assert(path);
+    assert(*path == '/');
+
+    int path_len = strlen(path);
+    char *r = path; /* end of so far constructed path */
+    int seg_len = 0;
+    const char *seg = r; /* segment to be appended to the path */
+    while (seg < path + path_len) {
+        seg_len = 0;
+        while (*seg && *seg == '/') {
+            seg++; /* skip slashes */
+        }
+        while (seg[seg_len] && seg[seg_len] != '/') {
+            seg_len++; /* length of segment to be copied */
+        }
+        if (*seg && r + seg_len + 1 <= path + path_len) {
+            memmove(r + 1, seg, seg_len);
+            r = r + seg_len + 1;
+            *r = seg[seg_len]; /* either '\0' or '/' */
+        }
+        seg += seg_len;
+    }
+    if (*r) {
+        *++r = '\0';
+    }
+}
+
+int vfs_unlink_recursive(const char *root, char *path_buf, size_t max_size)
+{
+    assert(root);
+    assert(path_buf);
+
+    /* This function works like a Depth-first search (DFS).
+       First, we go as deep as we can into a directory and delete contained files.
+       Then we delete the now empty directory and go to the parent directory
+       and repeat the process. */
+    int err;
+    if (*root != '/' || !strcmp(root, "/")) {
+        /* only accept absolute paths and not the FS root */
+        return -EINVAL;
+    }
+    if (strlen(root) >= max_size) {
+        return -ENOBUFS;
+    }
+    strcpy(path_buf, root);
+    _vfs_prepare_path(path_buf);
+    if (path_buf[strlen(path_buf) - 1] != '/') {
+        if ((err = vfs_is_dir(path_buf)) < 0) {
+            return err; /* early unexpected error */
+        }
+        else if (!err) {
+            /* just a file */
+            return vfs_unlink(path_buf);
+        }
+        strcat(path_buf, "/");
+    }
+    vfs_DIR dir;
+    vfs_dirent_t entry;
+    char seg[VFS_NAME_MAX + 1] = {0}; /* + 1 to append a '/' */
+    size_t seg_len, root_len, fin = strlen(path_buf);
+    while ((root_len = strlen(path_buf)) >= fin) {
+        strcat(path_buf, seg);
+        if ((err = vfs_opendir(&dir, path_buf)) < 0) { /* this works with a trailing '/' */
+            return err;
+        }
+        while (vfs_readdir(&dir, &entry) > 0) {
+            if (!strncmp(entry.d_name, "..", (seg_len = strlen(entry.d_name)))) {
```

Comment:
No need to be so dense here - and since ".." is const anyway, no need for the `strncmp`

```suggestion
            seg_len = strlen(entry.d_name);
            if (!strcmp(entry.d_name, "..")) {
```